File: web/scripts/send-test-email.ts
#!/usr/bin/env tsx

/**
 * Test Email Sender for aclue Platform
 *
 * This script sends test emails using the React Email templates to verify
 * how emails appear in real email clients. Uses Resend for reliable delivery
 * and includes comprehensive error handling and validation.
 *
 * Features:
 * - TypeScript implementation with proper error handling
 * - Real email sending via Resend API
 * - Multiple test scenarios and sample data
 * - Configuration validation and safety checks
 * - British English throughout all communications
 *
 * Usage:
 *   npm run email:send-test <recipient@example.com>
 *   or directly: npx tsx scripts/send-test-email.ts <recipient@example.com>
 *
 * Environment Variables Required:
 *   RESEND_API_KEY - Your Resend API key
 *   FROM_EMAIL - Sender email address (optional, defaults to noreply@aclue.app)
 */

import { Resend } from 'resend';
import { render } from '@react-email/render';
import { createElement } from 'react';
import { readFileSync } from 'fs';
import { resolve } from 'path';

// Import our React Email component
import WelcomeEmail from '../src/components/emails/WelcomeEmail';

interface EmailTestConfig {
  recipientEmail: string;
  fromEmail: string;
  resendApiKey: string;
  testMode: boolean;
}

interface TestEmailResult {
  success: boolean;
  messageId?: string;
  error?: string;
  timestamp: string;
}

class TestEmailSender {
  private resend: Resend;
  private config: EmailTestConfig;

  constructor(config: EmailTestConfig) {
    this.config = config;
    this.resend = new Resend(config.resendApiKey);
  }

  /**
   * Send welcome email test
   */
  async sendWelcomeEmailTest(): Promise<TestEmailResult> {
    console.log('📧 Sending welcome email test...');

    try {
      // Create React element with test data
      const emailElement = createElement(WelcomeEmail, {
        email: this.config.recipientEmail,
        source: 'email_test_script'
      });

      // Render HTML and text versions
      const htmlContent = await render(emailElement, { pretty: true });
      const textContent = await render(emailElement, { plainText: true });

      const subject = 'Welcome to aclue - AI-Powered Gift Discovery Awaits! 🎁 [TEST EMAIL]';

      console.log('📝 Email details:');
      console.log(`   To: ${this.config.recipientEmail}`);
      console.log(`   From: ${this.config.fromEmail}`);
      console.log(`   Subject: ${subject}`);
      console.log(`   HTML Size: ${Math.round(htmlContent.length / 1024)}KB`);
      console.log(`   Text Size: ${Math.round(textContent.length / 1024)}KB`);

      if (this.config.testMode) {
        console.log('🧪 TEST MODE: Email would be sent with the above configuration');
        return {
          success: true,
          messageId: 'test-mode-' + Date.now(),
          timestamp: new Date().toISOString()
        };
      }

      // Send via Resend
      const result = await this.resend.emails.send({
        from: this.config.fromEmail,
        to: [this.config.recipientEmail],
        subject,
        html: htmlContent,
        text: textContent,
        tags: [
          { name: 'type', value: 'welcome_email' },
          { name: 'source', value: 'test_script' },
          { name: 'environment', value: process.env.NODE_ENV || 'development' }
        ]
      });

      if (result.error) {
        throw new Error(`Resend API error: ${result.error.message}`);
      }

      console.log('✅ Welcome email sent successfully!');
      console.log(`📬 Message ID: ${result.data?.id}`);

      return {
        success: true,
        messageId: result.data?.id,
        timestamp: new Date().toISOString()
      };

    } catch (error) {
      console.error('❌ Failed to send welcome email:', error);
      return {
        success: false,
        error: error instanceof Error ? error.message : String(error),
        timestamp: new Date().toISOString()
      };
    }
  }

  /**
   * Send multiple test emails with different configurations
   */
  async sendTestSuite(): Promise<TestEmailResult[]> {
    console.log('🧪 Running complete email test suite...');
    console.log('=' .repeat(50));

    const results: TestEmailResult[] = [];

    // Test different email sources
    const testSources = ['newsletter', 'maintenance_page', 'landing_page'];

    for (const source of testSources) {
      console.log(`\n📧 Testing welcome email with source: ${source}`);

      try {
        const emailElement = createElement(WelcomeEmail, {
          email: this.config.recipientEmail,
          source
        });

        const htmlContent = await render(emailElement, { pretty: true });
        const textContent = await render(emailElement, { plainText: true });
        const subject = `aclue Welcome Email Test - Source: ${source} [TEST]`;

        console.log(`   Subject: ${subject}`);
        console.log(`   Source: ${source}`);

        if (this.config.testMode) {
          console.log('   🧪 TEST MODE: Email rendered successfully');
          results.push({
            success: true,
            messageId: `test-${source}-${Date.now()}`,
            timestamp: new Date().toISOString()
          });
          continue;
        }

        const result = await this.resend.emails.send({
          from: this.config.fromEmail,
          to: [this.config.recipientEmail],
          subject,
          html: htmlContent,
          text: textContent,
          tags: [
            { name: 'type', value: 'welcome_email_test' },
            { name: 'source', value: source },
            { name: 'test_suite', value: 'true' }
          ]
        });

        if (result.error) {
          throw new Error(`Resend API error: ${result.error.message}`);
        }

        console.log(`   ✅ Sent successfully! Message ID: ${result.data?.id}`);

        results.push({
          success: true,
          messageId: result.data?.id,
          timestamp: new Date().toISOString()
        });

        // Add delay between emails to avoid rate limiting
        await new Promise(resolve => setTimeout(resolve, 1000));

      } catch (error) {
        console.error(`   ❌ Failed to send test for source ${source}:`, error);
        results.push({
          success: false,
          error: error instanceof Error ? error.message : String(error),
          timestamp: new Date().toISOString()
        });
      }
    }

    return results;
  }

  /**
   * Validate email configuration
   */
  static validateConfiguration(): EmailTestConfig | null {
    console.log('🔍 Validating email configuration...');

    const resendApiKey = process.env.RESEND_API_KEY;
    if (!resendApiKey) {
      console.error('❌ RESEND_API_KEY environment variable is required');
      console.error('   Please set your Resend API key in your .env file');
      return null;
    }

    if (!resendApiKey.startsWith('re_')) {
      console.error('❌ Invalid RESEND_API_KEY format');
      console.error('   Resend API keys should start with "re_"');
      return null;
    }

    const fromEmail = process.env.FROM_EMAIL || 'aclue <noreply@aclue.app>';

    console.log('✅ Configuration validated');
    console.log(`   Resend API Key: ${resendApiKey.substring(0, 10)}...`);
    console.log(`   From Email: ${fromEmail}`);

    return {
      recipientEmail: '', // Will be set by command line argument
      fromEmail,
      resendApiKey,
      testMode: process.env.NODE_ENV === 'test' || process.argv.includes('--test-mode')
    };
  }

  /**
   * Validate email address format
   */
  static isValidEmail(email: string): boolean {
    const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
    return emailRegex.test(email);
  }
}

/**
 * Display usage information
 */
function displayUsage(): void {
  console.log(`
📧 aclue Test Email Sender

Usage:
  npm run email:send-test <recipient@example.com> [options]
  npx tsx scripts/send-test-email.ts <recipient@example.com> [options]

Options:
  --test-mode    Don't send actual emails, just validate rendering
  --suite        Send complete test suite with multiple scenarios

Examples:
  npm run email:send-test john.doe@example.com
  npm run email:send-test test@aclue.app --suite
  npm run email:send-test demo@example.com --test-mode

Environment Variables Required:
  RESEND_API_KEY  Your Resend API key
  FROM_EMAIL      Sender email (optional, defaults to noreply@aclue.app)

For more information, visit: https://docs.aclue.app/email-testing
  `);
}

/**
 * Main function
 */
async function main(): Promise<void> {
  console.log('🚀 aclue Email Test System');
  console.log('=' .repeat(50));

  // Parse command line arguments
  const args = process.argv.slice(2);

  if (args.length === 0 || args.includes('--help') || args.includes('-h')) {
    displayUsage();
    return;
  }

  const recipientEmail = args[0];
  const runSuite = args.includes('--suite');
  const testMode = args.includes('--test-mode');

  // Validate recipient email
  if (!TestEmailSender.isValidEmail(recipientEmail)) {
    console.error('❌ Invalid email address format');
    console.error(`   Provided: ${recipientEmail}`);
    console.error('   Expected format: user@domain.com');
    process.exit(1);
  }

  // Validate configuration
  const config = TestEmailSender.validateConfiguration();
  if (!config) {
    console.error('❌ Configuration validation failed');
    process.exit(1);
  }

  config.recipientEmail = recipientEmail;
  config.testMode = config.testMode || testMode;

  if (config.testMode) {
    console.log('🧪 Running in TEST MODE - no emails will be sent');
  }

  // Create test sender
  const sender = new TestEmailSender(config);

  try {
    if (runSuite) {
      console.log(`\n📬 Running test suite for: ${recipientEmail}`);
      const results = await sender.sendTestSuite();

      console.log('\n📊 Test Suite Results:');
      console.log('=' .repeat(30));

      let successCount = 0;
      results.forEach((result, index) => {
        if (result.success) {
          console.log(`✅ Test ${index + 1}: Success (${result.messageId})`);
          successCount++;
        } else {
          console.log(`❌ Test ${index + 1}: Failed (${result.error})`);
        }
      });

      console.log(`\n📈 Summary: ${successCount}/${results.length} tests passed`);

      if (successCount === results.length) {
        console.log('🎉 All tests passed successfully!');
      } else {
        console.log('⚠️  Some tests failed - check the logs above');
        process.exit(1);
      }

    } else {
      console.log(`\n📬 Sending single welcome email test to: ${recipientEmail}`);
      const result = await sender.sendWelcomeEmailTest();

      if (result.success) {
        console.log('\n🎉 Email test completed successfully!');
        if (result.messageId) {
          console.log(`📬 Message ID: ${result.messageId}`);
        }
        if (!config.testMode) {
          console.log('📱 Check your email inbox to see how the email appears');
          console.log('💡 Test both light and dark mode if your email client supports it');
        }
      } else {
        console.log('\n❌ Email test failed');
        console.error(`Error: ${result.error}`);
        process.exit(1);
      }
    }

  } catch (error) {
    console.error('\n❌ Unexpected error during email testing:', error);
    process.exit(1);
  }
}

// Run the script if called directly
if (require.main === module) {
  main().catch(error => {
    console.error('Fatal error:', error);
    process.exit(1);
  });
}

export default TestEmailSender;

File: backend/scripts/send_test_email.py
#!/usr/bin/env python3
"""
Test Email Sender for Aclue Email Templates

This script allows developers to send actual test emails to preview how the templates
appear in real email clients. It uses the production email service but with test data.

Features:
- Send welcome email test
- Send admin notification test
- Support for different test scenarios
- Email validation and safety checks
- Environment configuration validation

Usage:
    python scripts/send_test_email.py --email your.email@example.com --type welcome
    python scripts/send_test_email.py --email admin@example.com --type admin
    python scripts/send_test_email.py --email test@example.com --type both

Safety Features:
- Only sends to explicitly provided email addresses
- Requires confirmation for sending
- Uses test data to avoid confusion
- Validates email service configuration
"""

import os
import sys
import asyncio
import argparse
from datetime import datetime
from pathlib import Path

# Add the parent directory to the path so we can import our modules
sys.path.append(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

from app.services.email_service import EmailService
from app.core.config import settings

class TestEmailSender:
    """
    Sends test emails using the production email service with test data.
    """

    def __init__(self):
        self.email_service = EmailService()

    async def send_welcome_email_test(self, test_email: str) -> bool:
        """Send a test welcome email"""
        print(f"📧 Sending welcome email test to: {test_email}")

        # Add test identifier to email content by temporarily modifying the source
        success = await self.email_service.send_welcome_email(
            email=test_email,
            source="email_preview_test"
        )

        if success:
            print(f"✅ Welcome email sent successfully to {test_email}")
        else:
            print(f"❌ Failed to send welcome email to {test_email}")

        return success

    async def send_admin_notification_test(self, test_email: str) -> bool:
        """Send a test admin notification email"""
        print(f"🔔 Sending admin notification test to: {test_email}")

        # Generate test data
        test_subscriber_email = "test.subscriber@preview.example.com"
        test_source = "email_preview_test"
        test_signup_id = f"preview-test-{datetime.now().strftime('%Y%m%d-%H%M%S')}"

        # Temporarily override admin email for testing
        original_admin_email = self.email_service.admin_email
        self.email_service.admin_email = test_email

        try:
            success = await self.email_service.send_admin_notification(
                subscriber_email=test_subscriber_email,
                source=test_source,
                signup_id=test_signup_id
            )

            if success:
                print(f"✅ Admin notification sent successfully to {test_email}")
            else:
                print(f"❌ Failed to send admin notification to {test_email}")

            return success

        finally:
            # Restore original admin email
            self.email_service.admin_email = original_admin_email

    def validate_configuration(self) -> bool:
        """Validate email service configuration"""
        print("🔧 Validating email service configuration...")

        issues = []

        # Check Resend API key
        if not settings.RESEND_API_KEY:
            issues.append("❌ RESEND_API_KEY not configured")
        else:
            print("✅ RESEND_API_KEY configured")

        # Check from email
        if not self.email_service.from_email:
            issues.append("❌ From email not configured")
        else:
            print(f"✅ From email: {self.email_service.from_email}")

        # Check admin email
        if not self.email_service.admin_email:
            issues.append("❌ Admin email not configured")
        else:
            print(f"✅ Admin email: {self.email_service.admin_email}")

        if issues:
            print("\n🚨 Configuration Issues:")
            for issue in issues:
                print(f"   {issue}")
            return False

        print("✅ All configuration checks passed!")
        return True

    def validate_email_address(self, email: str) -> bool:
        """Basic email address validation"""
        import re
        pattern = r'^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$'
        return re.match(pattern, email) is not None

def main():
    """Main function for sending test emails"""
    parser = argparse.ArgumentParser(
        description="Send test emails using Aclue email templates",
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog="""
Examples:
  python scripts/send_test_email.py --email john@example.com --type welcome
  python scripts/send_test_email.py --email admin@company.com --type admin
  python scripts/send_test_email.py --email test@domain.com --type both
  python scripts/send_test_email.py --check-config

Safety Notes:
  - This script sends real emails to the specified address
  - Test emails include clear identification as test messages
  - Only use with email addresses you control
  - Requires valid Resend API configuration
        """
    )

    parser.add_argument(
        '--email',
        type=str,
        help='Email address to send test emails to'
    )

    parser.add_argument(
        '--type',
        choices=['welcome', 'admin', 'both'],
        default='welcome',
        help='Type of test email to send (default: welcome)'
    )

    parser.add_argument(
        '--check-config',
        action='store_true',
        help='Only check email service configuration without sending emails'
    )

    parser.add_argument(
        '--confirm',
        action='store_true',
        help='Skip confirmation prompts (use with caution)'
    )

    args = parser.parse_args()

    # Initialize the test email sender
    sender = TestEmailSender()

    # Validate configuration first
    if not sender.validate_configuration():
        print("\n❌ Email service configuration is invalid. Please check your environment variables.")
        return False

    # If only checking configuration, exit here
    if args.check_config:
        print("\n✅ Configuration check completed successfully!")
        return True

    # Validate email argument
    if not args.email:
        print("❌ Email address is required. Use --email your.address@example.com")
        return False

    if not sender.validate_email_address(args.email):
        print(f"❌ Invalid email address format: {args.email}")
        return False

    # Safety confirmation
    if not args.confirm:
        print(f"\n⚠️  WARNING: This will send real emails to {args.email}")
        print(f"📧 Email type: {args.type}")
        print(f"🕐 Timestamp: {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}")

        confirm = input("\nDo you want to proceed? (type 'yes' to continue): ").strip().lower()
        if confirm != 'yes':
            print("❌ Test email sending cancelled by user.")
            return False

    # Send test emails
    async def send_tests():
        success_count = 0
        total_count = 0

        if args.type in ['welcome', 'both']:
            total_count += 1
            if await sender.send_welcome_email_test(args.email):
                success_count += 1

        if args.type in ['admin', 'both']:
            total_count += 1
            if await sender.send_admin_notification_test(args.email):
                success_count += 1

        return success_count, total_count

    try:
        print(f"\n🚀 Sending test emails...")
        success_count, total_count = asyncio.run(send_tests())

        print(f"\n📊 Test Email Results:")
        print(f"   ✅ Successful: {success_count}/{total_count}")
        print(f"   📧 Sent to: {args.email}")
        print(f"   🕐 Completed: {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}")

        if success_count == total_count:
            print(f"\n🎉 All test emails sent successfully!")
            print(f"📱 Check your email client ({args.email}) to see how the templates appear.")
            return True
        else:
            print(f"\n⚠️  Some test emails failed to send. Check the logs above for details.")
            return False

    except Exception as e:
        print(f"\n❌ Error sending test emails: {str(e)}")
        return False

if __name__ == "__main__":
    success = main()
    sys.exit(0 if success else 1)